vitrivr-engine-module-torchserve/src/main/kotlin/org/vitrivr/engine/features/external/torchserve/basic/TorchServe.kt

```diff
@@ -0,0 +1,139 @@
+package org.vitrivr.engine.features.external.torchserve.basic
+
+import com.google.protobuf.ByteString
+import org.vitrivr.engine.core.context.IndexContext
+import org.vitrivr.engine.core.features.averagecolor.AverageColor
+import org.vitrivr.engine.core.model.content.element.ContentElement
+import org.vitrivr.engine.core.model.content.element.ImageContent
+import org.vitrivr.engine.core.model.descriptor.Descriptor
+import org.vitrivr.engine.core.model.descriptor.vector.FloatVectorDescriptor
+import org.vitrivr.engine.core.model.metamodel.Analyser
+import org.vitrivr.engine.core.model.metamodel.Schema
+import org.vitrivr.engine.core.model.retrievable.Retrievable
+import org.vitrivr.engine.core.operators.Operator
+import org.vitrivr.engine.module.torchserve.client.InferenceClient
+
+/**
+ * An abstract implementation of the [TorchServe] [Analyser], which leverages TorchServe ML models for inference.
+ *
+ * @author Ralph Gasser
+ * @version 1.0.0
+ */
+abstract class TorchServe<C : ContentElement<*>, D : Descriptor<*>> : Analyser<C, D> {
+
+    companion object {
+        /** Key for the TorchServe 'host' parameter. */
+        const val TORCHSERVE_HOST_KEY = "host"
+
+        /** Default for the TorchServe 'host' parameter. */
+        const val TORCHSERVE_HOST_DEFAULT = "127.0.0.1"
+
+        /** Key for the TorchServe 'port' parameter. */
+        const val TORCHSERVE_PORT_KEY = "port"
+
+        /** Default value for the TorchServe 'port' parameter. */
+        const val TORCHSERVE_PORT_DEFAULT = 7070
+
+        /** Key for the TorchServe 'token' parameter. */
+        const val TORCHSERVE_TOKEN_KEY = "token"
+
+        /** Key for the TorchServe 'model' parameter. */
+        const val TORCHSERVE_MODEL_KEY = "model"
+
+    }
+
+    /** A cached version of the [InferenceClient]. */
+    @Volatile
+    private var cachedClient: InferenceClient? = null
+
+    /**
+     * Performs the [AverageColor] analysis on the provided [List] of [ImageContent] elements.
+     *
+     * @param content The [List] of [ImageContent] elements.
+     * @return [List] of [FloatVectorDescriptor]s.
+     */
+    fun analyse(content: Collection<C>, model: String, host: String, port: Int = 8080, token: String? = null): List<D> {
+        /* Obtain a client. */
+        var client = synchronized(this) {
+            var localClient = this.cachedClient
+            if (localClient == null || localClient.host != host || localClient.port != port) {
+                localClient = InferenceClient(host, port, token)
+                this.cachedClient?.close()
+                this.cachedClient = localClient
+            }
+            localClient
+        }
+
+        val descriptors = mutableListOf<D>()
+        for (c in content) {
+            /* Convert the content to a ByteString. */
+            val data = this.toByteString(c)
```

Comment:
I think there is this fundamental question (which keeps popping up for me): Does a method like this act on a single `ContentElement` or a `Retrievable`. I guess for the latter it makes sense to return a map, whereas there is no way to add an additional parameter for the former.

There are architectual consequences to the two choices. Most importantly, if I simply hand the `Retirevable` to the method, it must be able to decide which `ContentElement`s to consider.

Reply:
Oh and: Since this method might be used during retrieval as well, we would probably need both anyway since in this scenario we don't get a `Retrievable`.

Reply:
I'd argue we need the `ContentElement` to map logic in any case and more specialized functionality can be added later.